desc/protoparse/ast/file.go

```diff
@@ -0,0 +1,230 @@
+package ast
+
+import "fmt"
+
+// FileNode is the root of the AST hierarchy.
+type FileNode struct {
+	compositeNode
+	Syntax   *SyntaxNode // nil if file has no syntax declaration
+
+	// Any package declarations in the file. Note that a valid file
+	// will have only zero or one such declaration.
+	Package  []*PackageNode
+	Imports  []*ImportNode
+	Options  []*OptionNode
+	Messages []*MessageNode
+	Enums    []*EnumNode
+	Extends  []*ExtendNode
+	Services []*ServiceNode
+
+	// All of the above elements in their original order as encountered
+	// in the source file.
+	AllDecls    []FileElement
+
+	// This provides access to all comments in the entire file.
+	AllComments []Comment
```

Comment:
Is this just top-level comments, or recursive comments on each element?

Reply:
The latter -- it is a reference to every comment in the file.

Reply:
I added this just because I noticed `go/ast.File` does something similar.

Reply:
I decided to remove it. Since it is a `[]Comment`, not a `[]*Comment`, it doesn't useful (since you can't modify the comments or their metadata for example). Maybe I'll come back and add a `[]*Comment` field for this later, but I'll leave it out for now, until it's obviously useful...